sw/applications/example_spi_host_quadIO/main.c

```diff
@@ -0,0 +1,309 @@
+        soc_ctrl_select_spi_host(&soc_ctrl);
+    #endif
+
+
+    // Enable SPI host device
+    spi_set_enable(&spi_host, true);
+
+    // Enable event interrupt
+    spi_enable_evt_intr(&spi_host, true);
+    // Enable RX watermark interrupt
+    spi_enable_rxwm_intr(&spi_host, true);
+    // Enable SPI output
+    spi_output_enable(&spi_host, true);
+
+    // Configure SPI clock
+    // SPI clk freq = 1/2 core clk freq when clk_div = 0
+    // SPI_CLK = CORE_CLK/(2 + 2 * CLK_DIV) <= CLK_MAX => CLK_DIV > (CORE_CLK/CLK_MAX - 2)/2
+    uint16_t clk_div = 0;
+    if(FLASH_CLK_MAX_HZ < core_clk/2){
+        clk_div = (core_clk/(FLASH_CLK_MAX_HZ) - 2)/2; // The value is truncated
+        if (core_clk/(2 + 2 * clk_div) > FLASH_CLK_MAX_HZ) clk_div += 1; // Adjust if the truncation was not 0
+    }
+    // SPI Configuration
+    // Configure chip 0 (flash memory)
+    const uint32_t chip_cfg = spi_create_configopts((spi_configopts_t){
+        .clkdiv     = clk_div,
+        .csnidle    = 0xF,
+        .csntrail   = 0xF,
+        .csnlead    = 0xF,
+        .fullcyc    = false,
+        .cpha       = 0,
+        .cpol       = 0
+    });
+    spi_set_configopts(&spi_host, 0, chip_cfg);
+    spi_set_csid(&spi_host, 0);
+
+    // Set RX watermark to 8 word
+    spi_set_rx_watermark(&spi_host, 8);
+
+    uint32_t *flash_data_ptr = flash_data[0];
+
+
+    // ----------------COMMAND----------------
+    // Power up flash
+    // ----------------COMMAND----------------
+    
+    // Create segment 1
+    const uint32_t powerup_byte_cmd = W25Q128JW_CMD_RELEASE_POWERDOWN;
+    spi_write_word(&spi_host, powerup_byte_cmd);
+
+    const uint32_t cmd_powerup = spi_create_command((spi_command_t){
+        .len        = 0,                 // 1 Byte
+        .csaat      = false,             // End command
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirTxOnly      // Write only
+    });
+    spi_set_command(&spi_host, cmd_powerup);
+    spi_wait_for_ready(&spi_host);
+    // ----------------END COMMAND----------------
+
+
+    volatile uint32_t data_addr = flash_original;
+    
+
+    // ----------------COMMAND----------------
+    // Fast Read Quad I/O
+    // ----------------COMMAND----------------
+
+    // Create segment 1
+    uint32_t cmd_read_quadIO = W25Q128JW_CMD_READ_QUADIO;
+    spi_write_word(&spi_host, cmd_read_quadIO);
+    spi_wait_for_ready(&spi_host);
+
+    const uint32_t cmd_read = spi_create_command((spi_command_t){
+        .len        = 0,                 // 1 Byte
+        .csaat      = true,              // Command not finished
+        .speed      = kSpiSpeedStandard, // Single speed
+        .direction  = kSpiDirTxOnly      // Write only
+    });
+    spi_set_command(&spi_host, cmd_read);
+    spi_wait_for_ready(&spi_host);
+
+
+    // Create segment 2
+    uint32_t read_byte_cmd = (REVERT_24b_ADDR(flash_original) | 0xFF << 24); // Fxh (here FFh) required by W25Q128JW
```

Comment:
why?

Reply:
[W25Q128JW datasheet](https://datasheet.lcsc.com/lcsc/2103261540_Winbond-Elec-W25Q128JWSIQ_C2763561.pdf) on page 33 describe the behaviour of line 221.
In short, after the command code at single speed the flash expect to receive the memory address MSB first at quad speed(so a Byte swap is require in a little-endian architecture). After it, a dummy byte at quad speed of value Fxh is required.
A detailed showcase of this behaviour can be also found in one [OpenTitan diagram](https://opentitan.org/book/hw/ip/spi_host/doc/programmers_guide.html#byte-ordering-conventions).

Reply:
sorry I was not clear, what I meant this comment `Fxh (here FFh) required by W25Q128JW` I do not understand what Fxh means

Reply:
I found no clear description on the data sheet. I though it is an hexadecimal with first 4 bits all ones (F) and the next for can be whatever (x).
Anyway in the opentitan documentation (linked in the message before) I saw a diagram with the byte after the address with all ones so I decided to implement it like opentitan (FFh).

Reply:
ok thanks for the clarification

Reply:
another question @consanii , I see in the single mode SPI examples that the address is shifted by 8 after reversed, and then ored with `03`, here you do reverse, or with `ff`, but I do not see the shift by 8, why?

Reply:
Because the behaviour in single and quad modes differs. In quad mode, the command code must be sent at standard speed before sending the address + Fxh afterward. There is no need to shift the address to make space for the command because the command is already transmitted.